core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -815,6 +816,25 @@ class Partition(val topicPartition: TopicPartition,
       case None => localReplica.logEndOffset.messageOffset
     }
 
+    // Only actually check the HW if this is a client request and _not_ while unclean leader
+    // election is enabled
+    val shouldCheckHW = {
+      !logManager.currentDefaultConfig.uncleanLeaderElectionEnable &&
```

Comment:
I think we'd want to do this validation even if unclean leader election is enabled. The point that the KIP was making is that high watermark monotonicity _can_ be violated if unclean leader election is enabled. But we'd still want to minimize such violations to cases when an unclean leader election has actually happened.